packages/flutter/lib/src/foundation/basic_types.dart

```diff
@@ -280,3 +280,37 @@ class _LazyListIterator<E> implements Iterator<E> {
     return true;
   }
 }
+
+/// A factory interface, that also reports the type of the created objects.
+///
+/// The equality of [Factory] object is determined solely by [Factory.type].
+class Factory<T> {
+  /// Creates a new factory.
+  ///
+  /// If `type` is null, [Factory.type] returns [T], else it returns `type`.
+  ///
+  /// The `constructor` parameter must not be null.
+  const Factory(this.constructor, {Type type})
+      : assert(constructor != null),
+        _type = type;
+
+  /// Creates a new object of type T.
+  final ValueGetter<T> constructor;
+
+  final Type _type;
+  /// The type of the objects created by this factory.
+  Type get type => _type ?? T;
+
+  @override
+  bool operator ==(Object other) {
+    if (other is! Factory<T>) {
+      return false;
+    }
+    final Factory<T> typedOther = other;
+    return type == typedOther.type;
+  }
```

Comment:
it seems weird that two factories with different constructor would be `==`. Maybe have a separate method that compares factories for equivalence?

Reply:
(oh but i guess you need this so that they match in a set?)

Reply:
this is a tough one, but i think the right call here is to use a map indexed by type, and to have the factories either not have a custom ==, or to have their custom == also compare the actual constructor.

I'm worried about the confusion that one day someone would have if they did:

```dart
  var a = Factory(() => 'a');
  var b = Factory(() => 'b');
  print(a == b); // true (!)
```

Reply:
ugh, but you really want it to be a set for the rest of the API...

Reply:
maybe the right solution here is to make this not be a general-purpose class and just have a class specific to this purpose, with a more specific name...

Reply:
or alternatively, still take a set, but then independently verify that the set only has one of each type, and match on type when the new set comes in?